### ISSUE TITLE
Mock ansible invocation in unit test

### ISSUE THREAD

**Issue by gciavarrini** [project-flotta/flotta-device-worker#123]
At the moment `ansible` must be installed to successfully run unit tests of `package ansible` because they rely on real invocations of `ansible-playbook` command (via `go-ansible` library).

Unit-test shouldn't depend on external resources.

We need to mock them or to present an API that will encapsulate the actual call to the external service.

**Comment by gabriel-farache** [comment]
Are tests in ansible_manager_test.go really unit tests and not integrations tests?
The unit test would be to check if we go inside the HandlePlaybook() and ExecutePendingPlaybooks() that are supposed to execute the ansible playbook.
By checking the output of the execution of the playbook, it looks like an integration test
One easy way would be to run tests in docker container in which all external dependencies are installed for integrations tests: 

- So in the Makefile add a target to run integration tests that will executes the ginkgo in the docker. 
- The target should be able to take specific folders to run tests on in case the user does not want to run all tests but only a subset

What do you think?

**Comment by gciavarrini** [comment]
You're right: they are integration tests. The goals of the issue is to move them in the integration test section (maybe we can reword the issue description).

Regarding implementation I think @machacekondra can help :)

**Comment by gabriel-farache** [comment]
Ok so the goal of this issue is to create a new folder test/integration and to move the ansible tests in it, correct?
Then  add a target in the makefile to run those tests (see my above comment about how I would do it), @machacekondra any advices?

**Comment by gciavarrini** [comment]
The `test/e2e` folder already exists. 

For this issue we should mock ansible invocation and refactor the current unit tests to be "proper" unit test (https://issues.redhat.com/browse/ECOPROJECT-691).

We also planned to add [e2e CI test](https://issues.redhat.com/browse/ECOPROJECT-598) but, I think, It can be another issue/PR .

**Comment by gabriel-farache** [comment]
There is no e2e folder in the device-worker project in the main branch at least and the JIRA ticket has "flotta operator" as component :)
The current e2e tests are not executed in the CI currently? Because I see a target in the Makefile
So I am a little confused to what should be done here: move the tests to correct folder and maybe create a Maefile target to run tests in docker container (at least for ansible tests) to avoid being dependent of the platform, or move the test and then create unit tests?

**Comment by gciavarrini** [comment]
> There is no e2e folder in the device-worker project

AFAIK, the e2e test should be part of the operator. The folder `test\e2e` is there.

> the JIRA ticket has "flotta operator" as component :)

Fixed! Thanks :)

> So I am a little confused to what should be done here

Actual test needs to be part of e2e: I've no good suggestion on the implementation for you but I think they should be part of the operator, so we need to wait for the implementation step (we are still defining the CRDs).

IMO, for this specific issue the goal is to mock ansible invocation in the current available tests. This step it's necessary to make unit test independent from the availability of `ansible` command and to ease the writing of proper unit test. 

Having said that, if anyone has different opinions on this issue I'd like to hear them! :)

**Comment by gabriel-farache** [comment]
Yes, integrations tests should be in e2e folder, I agree. But in the device-worker project, such folder does not exists that why I suggested to create one in the project to put the current ansible tests that are integration tests in it.

As for mocking the AnsiblePlaybookCmd (as , from my understanding this is the playbookCmd that is used to used to execute the ansible playbook) in the dispatcher I agree that mocking it would be useful

Maybe we could discuss this at tomorrow team meeting if that is fine with you and if it fits in the agenda
